OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,293 @@
+/* -------------------------------------------------------------------------- *
+ *                            OpenSim:  DataTable.h                           *
+ * -------------------------------------------------------------------------- *
+ * The OpenSim API is a toolkit for musculoskeletal modeling and simulation.  *
+ * See http://opensim.stanford.edu and the NOTICE file for more information.  *
+ * OpenSim is developed at Stanford University and supported by the US        *
+ * National Institutes of Health (U54 GM072970, R24 HD065690) and by DARPA    *
+ * through the Warrior Web program.                                           *
+ *                                                                            *
+ * Copyright (c) 2005-2015 Stanford University and the Authors                *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#ifndef OPENSIM_DATATABLE_H_
+#define OPENSIM_DATATABLE_H_
+
+/** \file
+This file defines the  DataTable_ class, which is used by OpenSim to provide an 
+in-memory container for data access and manipulation.                         */
+
+// Non-standard headers.
+#include "SimTKcommon.h"
+#include "OpenSim/Common/Exception.h"
+#include "OpenSim/Common/ValueArrayDictionary.h"
+
+namespace OpenSim {
+
+/** AbstractDataTable is the base-class of all DataTable_(templated) allowing 
+storage of DataTable_ templated on different types to be stored in a container 
+like std::vector. AbstractDataTable_ offers:
+- Interface to access column metadata of the DataTable_.
+- A heterogeneous container to store metadata associated with the DataTable_ in
+  the form of key-value pairs where key is of type std::string and value can be
+  of any type.
+
+This class is abstract and cannot be used directly. Create instances of 
+DataTable_ instead. See DataTable_ for details on ussage.                     */
+class AbstractDataTable {
+protected:
+    /** ValueDictionary represents an associative array mapping from string to
+    an AbstractValue.                                                         */
+    class ValueDictionary {
+    public:
```

Comment:
> I believe a ValueDictionary will be a useful class outside the context of data tables, so we might consider it not being nested within AbstractDataTable. 

Do you think we'll need that in addition to ValueArrayDictionary? I think it is simpler just to have a single class since a one-element value array works just fine as a value. Also I think if both classes exist they should be related somehow so that if I have a (single-) ValueDictionary I can supply it where a (multiple-) ValueArrayDictionary is expected, since a one-element value is a kind of array value (with array length 1). This same issue arose with Property where lists of various lengths naturally occur, with lists of length 1 being a popular subset.